vertexarray.cpp

```diff
@@ -0,0 +1,51 @@
+#include "vertexarray.h"
+
+#include <vector>
+
+#include "renderutils.h"
+
+VertexArray::VertexArray() :
+    mRendererID(0)
+{
+    GLCall(glGenVertexArrays(1, &mRendererID));
+}
+
+VertexArray::~VertexArray()
+{
+    GLCall(glDeleteVertexArrays(1, &mRendererID));
+}
+
+void VertexArray::AddBuffer(const VertexBuffer& vb, const VertexBufferLayout& layout)
+{
+    Bind();
+    vb.Bind();
+
+    const std::vector<VertexBufferElement>& elements = layout.GetElements();
+    unsigned int offset = 0;
+    for (int i = 0; i < elements.size(); i++)
+    {
+        const VertexBufferElement element = elements[i];
+
+        GLCall(glEnableVertexAttribArray(i));
+
+// Ugly warning suppression.
+// C4312: 'type cast': conversion from 'unsigned int' to 'const void *' of greater size
+// unfortunately, this is the only way to do it. (afaik)
+#pragma warning(suppress: 4312)
+        GLCall(glVertexAttribPointer(i,
+            element.mCount, element.mType, element.mNormalized,
+            layout.GetStride(), (const void *)(offset)));
```

Comment:
This was a warning I was hitting. And because I want to keep it right from the start, I have /WX enabled to turn them into errors. I was not able to find a work around to doing this warning suppression. Even doing a suggested `reintepret_cast<const void *>` was also giving the same warning. Is there anything that can be done about this?

Reply:
I think you possibly have to do a multi step cast to ensure you aren't losing data if the type goes from 64->32 bit, something like (the extremely ugly)

```
GLCall(glVertexAttribPointer(i,
            element.mCount, element.mType, element.mNormalized,
            layout.GetStride(), reinterpret_cast<const void *>(static_cast<uintptr_t>(offset))));
```

Not sure which is better..

Reply:
Yeah that managed to fix it! Definitely an incantation that line is, but I think I would like to avoid warning suppression if I can.

Reply:
Yeah I think you are wise to avoid erro suppression. You could break it up and add a comment if you wanted:

```
// Have to do this crazy cast because reasons
auto offsetPointer = reinterpret_cast<const void *>(static_cast<uintptr_t>(offset));
```

This is also a great opportunity to wade a little bit into appreciating `auto`. The type is already on the right side of the equals sign, why not avoid the extra clutter? :)